src/axom/primal/tests/primal_nurbs_curve.cpp

```diff
@@ -0,0 +1,831 @@
+{
+  SLIC_INFO("Testing NURBS knot insertion");
+
+  const int DIM = 3;
+  using CoordType = double;
+  using PointType = primal::Point<CoordType, DIM>;
+  using NURBSCurveType = primal::NURBSCurve<CoordType, DIM>;
+
+  PointType data[4] = {PointType {0.6, 1.2, 1.0},
+                       PointType {1.3, 1.6, 1.8},
+                       PointType {2.9, 2.4, 2.3},
+                       PointType {3.2, 3.5, 3.0}};
+
+  double weights[4] = {1.0, 2.0, 3.0, 4.0};
+
+  NURBSCurveType curve(data, weights, 4, 3);
+  NURBSCurveType curve_knots(data, weights, 4, 3);
+
+  // Knot insertion shouldn't change the paramterization or position
+  //  of the curve.
+
+  // Insert a knot at 0.5
+  curve_knots.insertKnot(0.5, 3);
+  for(double t = 0.0; t <= 1.0; t += 0.1)
+  {
+    PointType p = curve.evaluate(t);
+    PointType p_knots = curve_knots.evaluate(t);
+    for(int i = 0; i < DIM; ++i)
+    {
+      EXPECT_NEAR(p[i], p_knots[i], 1e-13);
+    }
+  }
+
+  curve_knots.insertKnot(0.7, 1);
+  for(double t = 0.0; t <= 1.0; t += 0.1)
+  {
+    PointType p = curve.evaluate(t);
+    PointType p_knots = curve_knots.evaluate(t);
+    for(int i = 0; i < DIM; ++i)
+    {
+      EXPECT_NEAR(p[i], p_knots[i], 1e-13);
+    }
+  }
+
+  curve_knots.insertKnot(0.4, 2);
+  for(double t = 0.0; t <= 1.0; t += 0.1)
+  {
+    PointType p = curve.evaluate(t);
+    PointType p_knots = curve_knots.evaluate(t);
+    for(int i = 0; i < DIM; ++i)
+    {
+      EXPECT_NEAR(p[i], p_knots[i], 1e-13);
+    }
+  }
+
+  // Inserting knots shouldn't increase its multiplicty
+  //  greater than the degree
+  const auto num_knots = curve_knots.getNumKnots();
+  curve_knots.insertKnot(0.5, 3);
+  curve_knots.insertKnot(0.0, 1);
+  curve_knots.insertKnot(1.0, 1);
+  EXPECT_EQ(curve_knots.getNumKnots(), num_knots);
+
+  curve_knots.insertKnot(0.4, 2);
+  EXPECT_EQ(curve_knots.getNumKnots(), num_knots + 1);
+
+  for(double t = 0.0; t <= 1.0; t += 0.1)
+  {
+    PointType p = curve.evaluate(t);
+    PointType p_knots = curve_knots.evaluate(t);
+    for(int i = 0; i < DIM; ++i)
+    {
+      EXPECT_NEAR(p[i], p_knots[i], 1e-13);
+    }
+  }
+}
+
+//------------------------------------------------------------------------------
+TEST(primal_nurbscurve, curve_splitting)
+{
+  SLIC_INFO("Testing NURBS curve splitting");
+
+  const int DIM = 3;
+  using CoordType = double;
+  using PointType = primal::Point<CoordType, DIM>;
+  using NURBSCurveType = primal::NURBSCurve<CoordType, DIM>;
+
+  PointType data[4] = {PointType {0.6, 1.2, 1.0},
+                       PointType {1.3, 1.6, 1.8},
+                       PointType {2.9, 2.4, 2.3},
+                       PointType {3.2, 3.5, 3.0}};
+
+  double weights[4] = {1.0, 2.0, 3.0, 4.0};
+
+  for(int deg = 2; deg <= 2; ++deg)
+  {
+    NURBSCurveType curve(data, weights, 4, deg);
+
+    // Do some knot insertion to make it interesting
+    curve.insertKnot(0.3, 2);
+    curve.insertKnot(0.6, 1);
+    curve.insertKnot(0.8, 1);
```

Comment:
Should there be some checks here that the curve and or/or knots have the right shapes after each knot insertion.
(These could also be tested in separate unit tests).

Reply:
Some of this behavior is tested in the `insertKnot` test, but I've made it more thorough.